cmds/dhclient/dhclient.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// dhclient setups up DHCP.
+//
+// Synopsis:
+//     dhclient [OPTIONS...]
+//
+// Options:
+//     -timeout:  lease timeout in seconds
+//     -renewals: number of DHCP renewals before exiting
+//     -verbose:  verbose output
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"time"
+
+	"github.com/d2g/dhcp4"
+	"github.com/d2g/dhcp4client"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	defaultIface = "eth0"
+	// slop is the slop in our lease time.
+	slop = 10 * time.Second
+)
+
+var (
+	leasetimeout = flag.Int("timeout", 600, "Lease timeout in seconds")
+	renewals     = flag.Int("renewals", 2, "Number of DHCP renewals before exiting")
+	verbose      = flag.Bool("verbose", false, "Verbose output")
+	debug        = func(string, ...interface{}) {}
+)
+
+func dhclient(ifname string, timeout time.Duration, iList []string, done chan error) {
+	var err error
+
+	// if timeout is < 10 seconds, it's too short.
+	if timeout < slop {
+		timeout = 2 * slop
+		log.Printf("increased log timeout to %s", timeout)
+	}
+
+	n, err := ioutil.ReadFile(fmt.Sprintf("/sys/class/net/%s/address", ifname))
+	if err != nil {
+		done <- fmt.Errorf("cannot get mac for %v: %v", ifname, err)
+		return
+	}
+	// This is truly amazing but /sys appends newlines to all this data.
+	n = bytes.TrimSpace(n)
+	m, err := net.ParseMAC(string(n))
+	if err != nil {
+		done <- fmt.Errorf("mac error: %v", err)
+		return
+	}
+
+	iface, err := netlink.LinkByName(ifname)
+	if err != nil {
+		done <- fmt.Errorf("%s: netlink.LinkByName failed: %v", ifname, err)
+		return
+	}
+
+	c, err := dhcp4client.NewInetSock(dhcp4client.SetLocalAddr(net.UDPAddr{IP: net.IPv4(0, 0, 0, 0), Port: 68}), dhcp4client.SetRemoteAddr(net.UDPAddr{IP: net.IPv4bcast, Port: 67}))
+	if err != nil {
+		done <- fmt.Errorf("client conection generation: %v", err)
+		return
+	}
+
+	client, err := dhcp4client.New(dhcp4client.HardwareAddr(m), dhcp4client.Connection(c), dhcp4client.Timeout(timeout))
+	if err != nil {
+		done <- fmt.Errorf("error: %v", err)
+		return
+	}
+
+	// We require at least one successful request.
+	success, packet, err := client.Request()
+
+	if err != nil {
+		networkError, ok := err.(*net.OpError)
+		if ok && networkError.Timeout() {
+			done <- fmt.Errorf("%s: did not find a DHCP Server", n)
+			return
+		}
+		done <- fmt.Errorf("%s: Error:%v", n, err)
+		return
+	}
+
+	debug("Success on %s: %v\n", n, success)
+	debug("Packet: %v\n", packet)
+	debug("Lease is %v seconds\n", packet.Secs())
+
+	if !success {
+		done <- fmt.Errorf("we did not sucessfully get a DHCP Lease?")
+		return
+	}
+	log.Printf("IP Received: %v\n", packet.YIAddr().String())
+
+	for i := 0; i < *renewals; i++ {
+		addr := packet.YIAddr()
```

Comment:
from here until time.Sleep: split into a different function that just takes packet